_posts/2021-03-05-automatically-generated-api-tests-with-clojure-and-reitit.md

```diff
@@ -0,0 +1,332 @@
+                                              {:resource-id (:resource-id ctx)
+                                               :token       (:token ctx)}))}
+   "/api/resource"      {:post (test-path with-token {:content "some content"})}
+   "/api/resource/:resource-id"
+                        {:get    (test-path with-resource)
+                         :put    (test-path with-resource {:content "updated content"})
+                         :delete (test-path with-resource)}})
+```
+
+There are two parts for testing these paths. One is creating the necessary
+resources for testing, and the other is actually making the requests. Let's
+start off with the context. We need functions to create valid tokens, as well
+as some resources which can be updated. Here, `with-token` provides a context
+with a valid token, while `with-resource` provides a context with a pre-created
+resource.
+
+```clojure
+(def next-id (atom 0))
+
+(defn- create-token [ctx]
+  (assoc ctx
+    :token (swap! next-id inc)))
+
+(defn- create-resource [ctx]
+  (assoc ctx
+    :resource-id
+    (->
+      (request :post (:token ctx) "/api/resource" {:content "some content"})
+      :body
+      :resource-id)))
+
+(def empty-ctx (constantly {}))
+
+(defn with-token []
+  (-> (empty-ctx)
+      (create-token)))
+
+(defn with-resource []
+  (-> (empty-ctx)
+      (create-token)
+      (create-resource)))
+```
+
+What does `test-path` look like? Based on our usage, it needs to take in one or
+two arguments, the first of which is some function which creates the resources
+needed for testing, and the second is a body which will be sent to the API if
+provided. The function should also replace possible path parameters with
+appropriate id's from the context. In our use case, the only possible path
+parameter is `:resource-id`, but this function is easily extendable for
+additional parameters.
+
+```clojure
+(defn- create-request-path [path ctx]
+  (cond-> path
+          (:resource-id ctx) (string/replace ":resource-id" (-> ctx :resource-id str))))
+
+(defn test-path
+  ([context-fn] (test-path context-fn nil))
+  ([context-fn body]
+   (fn [path method]
+     (let [ctx          (context-fn)
+           request-path (create-request-path path ctx)
+           request-body (if (and (ifn? body) (not (map? body)))
+                          (body ctx)
+                          body)]
+       (testing "Request without tokens returns 403"
+         (is (= {:status 400 :body {:cause :missing-authentication}}
+                (request method nil request-path request-body))))
+       (testing "Request with tokens returns 200"
+         (is (= 200
+                (:status
+                  (request method (:token ctx) request-path request-body))))))
+     ; todo: add a test here which adds / removes the token from the resource
+   )))
+```
+
+With those two pieces of code, the  `test-all-paths` method goes over every
+path in the test-routes object, and creates API requests for the route.
```

Comment:
~method~ function

`test-routes` (backticks)

Maybe "map" instead of "object"?